relay/README.md

```diff
@@ -86,93 +64,144 @@ Scene 2:
 
-TODO: figure out forced addresses.
-TODO: figure out nested relayed connections.
+Double relay:
+
+- `/p2p-circuit/p2p/QmTwo/p2p-circuit/p2p/QmThree`
+  - Dial QmThree, through a relayed connection to QmTwo.
+  - The relay nodes will use peer routing to find an address for QmTwo and QmThree.
+  - We'll **not support nested relayed connections for now**, see [Future Work](#future-work) section.
 
 ## Wire format
 
-The wire format (or codec) is named `/ipfs/relay/circuit` and is simple.
-A variable-length header consisting of two length-prefixed multiaddrs
-is followed by a bidirectional stream of arbitrary data,
-and the eventual closing of the stream.
+We start the description of the Wire format by illustrating a possible flow scenario and then describing them in detail by phases.
+
+### Relay Message
+
+Every message in the relay protocol uses the following protobuf:
 
 ```
-<src><dst><data>
-
- ^    ^    ^
- |    |    |
- |    |    +-- bidirectional data stream
- |    |         (usually /multistream-select in the case of /p2p multiaddrs)
- |    |
- |    +------- multiaddr of the listening node
- |
- +------------ multiaddr of the dialing node
+message CircuitRelay {
+  enum Type { // RPC identifier, either HOP, STOP or STATUS
+    HOP = 1; 
+    STOP = 2; 
+    STATUS = 3;
+  }
+  
+  message Peer {
+    required bytes id = 1;    // peer id
+    repeated bytes addrs = 2; // peer's known addresses
+  }
+
+  optional Type type = 1;     // Type of the message
+
+  optional Peer srcPeer = 2;  // srcPeer and dstPeer are used when Type is HOP or STATUS
+  optional Peer dstPeer = 3;
+
+  optional String code = 4;   // Status code, used when Type is STATUS
+}
 ```
 
-After getting a stream to the relay node from its libp2p swarm,
-the dialing transport writes the header to the stream.
-The relaying node reads the header, gets a stream to the destination node,
-then writes the header to the destination stream and shortcircuits the two streams.
+### High level overview of establishing a relayed connection
+
+**Setup:**
+- Peers involved, A, B, R
+- A wants to connect to B, but needs to relay through R
+
+**Assumptions:**
+- A has connection to R, R has connection to B
+
+**Events:**
+- phase I: Open a request for a relayed stream (A to R).
+  - A dials a new stream `sAR` to R using protocol `/libp2p/circuit/relay/0.1.0`.
+  - A sends a CircuitRelay message with `{ type: 'HOP', srcPeer: '/p2p/QmA', dstPeer: '/p2p/QmB' }` to R through `sAR`.
+  - R receives stream `sAR` and reads the message from it.
+- phase II: Open a stream to be relayed (R to B).
+  - R opens a new stream `sRB` to B using protocol `/libp2p/circuit/relay/0.1.0`.
+  - R sends a CircuitRelay message with `{ type: 'STOP', srcPeer: '/p2p/QmA', dstPeer: '/p2p/QmB' }` on `sRB`.
+  - R sends a CircuitRelay message with `{ type: 'STATUS', code: 'OK' }` on `sAR`.
+- phase III: Streams are piped together, establishing a circuit
```

Comment:
@whyrusleeping I guess you're right. I remember dealing with some issues initially, which made me experiment a bit with we're the ok is sent, but most likely it was unrelated to the ordering of the messages. I'll give it another go sinse I'll have to rewrite message handling anyways.